README
== Bugzilla Lifecycles Calculator ==

This bit of software helps calculate the lifecycle of a bug or set of bugs. 

The lifecycle is the avg time in days passed for the bug to pass from state A to state B.

There are several kind of lifecycles calculated:

Overall: New --> Closed
Fixing:  New --> Resolved/Fixed
Releasing: Resolved/Fixed --> Released
Verification_fixed: Released/Fixed --> Verified
Verification_unfixed: Resolved/NotFixed --> Verified
Waiting: Waiting --> Resolved

The tool has 3 main routines:

1) Dump the DB content
2) Rebuild the history forward (bugzilla stores it backward)
3) Calculate lifecycles

Since the tool _has_ the history of all the bugs (with flags and keywords and stuff) it can
in theory be used to calculate any kind of statistics.

Also the tool has in the cfg file facilities to filter the bugs by type (severity and/or flags and/or keywords).

== Output ==

The scripts produce 2 kind of outputs: CSV and HTML.

HTML contains also the links to the bugtracking system instance.

CSV is mainly for post-processing.

The files end up in a dir that you need to configure in the cfg file


== Installation ==

Build the debian package and install it. Repackage it if you don't use Debian.


debian/changelog
bugzillametrics (0.4) unstable; urgency=low

  * Rebuild routine rewrote almost completely
    * Full history is now rebuilt taking care of when flags & keywords appeared
    * Tweaks to take care of status/priority modifications done in house
  * Fixing bugs in count phase where comments or flags on a resolved/fixed bug
    where interpreted as resolving/fixing date

 -- Stefano Mosconi <stefano.mosconi@nokia.com>  Tue, 30 Aug 2011 09:41:21 +0300

bugzillametrics (0.3) unstable; urgency=low

  * Getting command line options
  * Filters are now in configuration file

 -- Stefano Mosconi <stefano.mosconi@nokia.com>  Tue, 16 Aug 2011 17:36:33 +0300

bugzillametrics (0.2) unstable; urgency=low

  * new release of bugzillametrics 

 -- Stefano Mosconi <stefano.mosconi@nokia.com>  Fri, 22 Jul 2011 15:20:14 +0300

bugzillametrics (0.1) unstable; urgency=low

  * Initial release.

 -- Stefano Mosconi <stefano.mosconi@nokia.com>  Fri, 27 May 2011 10:58:45 +0300


test_rebuild.py
import tables
from datetime import datetime

debug = False

t = tables.openFile('/var/cache/bugzillametrics/cache.h5')
#for k in t.root.dumps.flags:
#    print 'bug_id:%s\nflags:%s\n' % (k['bug_id'],k['name'])
hist = t.root.history
hist_fields = hist.colnames
i = 0
n = len(hist)


bug_ids = list(frozenset([bug['id'] for bug in hist]))
bug_ids.sort()
#bug_ids = [250684]
print "Found %s bugs" % len(bug_ids)
chgd = 'changed_fields'
rem = 'removed'
additive_fields = ['flags', 'keywords']
broken_bugs = 0
broken_bugids = []
for bug_id in bug_ids:
    #print "========= BUG NUMBER %s ============" % bug_id
    bug_history = []
    #print '%s' % bug_id,end='\r'
    bug = []
    go = True
    while i < n and go:    
        #print hist[i]['id'], bug_id, end='\r'
        #print "%s" % i, end='\r'
        if hist[i]['id'] == bug_id:
            coord = dict(zip(hist_fields, hist[i])) 
            bug_history.append(coord)
        elif int(hist[i]['id']) > bug_id:
            go = False
        i += 1 
    bug_history.sort(cmp=lambda x,y:int(x['ts'])-int(y['ts']))
    #print frozenset([bug['id'] for bug in bug_history])
    broken = False
    # trying to rebuild the history
    for event in bug_history:
        if bug:
            # check what field changed
            changed = event[chgd].split('||')
            mutable_fields = [rem, chgd, 'ts']
            for field in changed:
                if field is not '':
                    curbroken = False
                    if event[rem] != '':
                        removed = dict([x.split('>>') for x in event[rem].split('||')])
                        if field not in additive_fields:
                            if removed[field] != bug[field]:
                                if debug:
                                    print "removed: %s  is different than %s" % (removed[field],bug[field])
                                broken = broken or True
                                curbroken = True
                        elif removed[field]:
                            bug_fields = set(bug[field].split(','))
                            rem_fields = set(removed[field].split(','))
                            if not rem_fields.issubset(bug_fields):
                                #pass
                                if debug:
                                    print "Rem fields is not a subset of fields", bug_fields, rem_fields
                                broken = broken or True
                                curbroken = True
                        else:
                            # There was a field changed but nothing was removed
                            # Make sure that really nothing was removed
                            cur_fields = set(event[field].split(','))
                            bug_fields = set(bug[field].split(','))
                            if not bug_fields.issubset(cur_fields) and bug_fields != set(['']):
                                if debug:
                                    print bug_fields
                                    print "Something got removed from %s" % field
                                    print datetime.fromtimestamp(bug['ts']),bug
                                    print "=========================="
                                    print datetime.fromtimestamp(event['ts']),event
                                broken = broken or True
                                curbroken = True
            # Now making sure that all the rest didn't change
            if changed:
                mutable_fields.extend(changed)
            for field in hist_fields:
                if field not in mutable_fields:
                    if bug[field] != event[field]:
                        if debug:
                            print "UPS... %s changed but it should have remained the same" % field
                            print datetime.fromtimestamp(bug['ts']),bug
                            print "=========================="
                            print datetime.fromtimestamp(event['ts']),event
                        broken = broken or True
                        curbroken = True
            # TODO: Last check would be to scan all the events in bugs_activity and make sure that 
            #       added and removed are present in history... but it's too long now...
            if curbroken and debug:
                print '\n\n'
                print "Bug #%s history is broken" % str(bug_id)
                print datetime.fromtimestamp(event['ts']),event

        bug = event.copy()
        #print datetime.fromtimestamp(bug['ts'])
        #print bug
        #print '========================='
    if broken:
        print 'history is not consistent for bug %s' % bug_id
        broken_bugs += 1
        broken_bugids.append(str(bug_id))


prc = float(broken_bugs)/float(len(bug_ids))*100

print 'Found %s broken bugs (%.2f%%)' % (broken_bugs, prc)
if broken_bugids:
    print 'Debug using --restrict=%s' % (','.join(broken_bugids))


bugzillametrics/dump.py
"""

"""
import logging

import sqlalchemy as sa


class Dump(object):
    """
    
    """
    
    def __init__(self, config, cache):
        """
        
        """
        self.config = config
        self.cache = cache
        bugs_restrictions = ba_restrictions = ''
        if self.config.run.restrict:
            r = self.config.run.restrict.split(':')
            if len(r) == 2:
                restr = 'BETWEEN %s AND %s' % (r[0],r[1])
            else:
                restr = 'in (%s)' % self.config.run.restrict
            bugs_restrictions = "AND id %s" % restr
            ba_restrictions = "AND a.bug_id %s" % restr
        
        # TODO: use classes
        self._tables = {
            'bugs': ['bug_id AS id', 'keywords', 'bug_status AS status', 'resolution',
                'bug_severity AS severity', 'UNIX_TIMESTAMP(delta_ts) AS ts', 'priority',
                'UNIX_TIMESTAMP(creation_ts) AS creation_ts'],
            'bugs_activity': ['a.bug_id AS bug_id', 'fieldid', 'added', 'removed', 'UNIX_TIMESTAMP(bug_when) AS ts', 'c.name AS program'],
            'classifications': ['id', 'name'],
        }

        self._tables_addons = {
            'bugs': {
                'prefix': ['SELECT * FROM ('],
                'cols': ['pd.name AS component', 'co.name AS subcomponent', 'c.name AS program'],
                'postfix': ['''LEFT JOIN products AS pd ON (a.product_id = pd.id)
                LEFT JOIN classifications AS c ON (pd.classification_id = c.id)
                LEFT JOIN components AS co ON (a.component_id = co.id)
                ORDER BY id, ts DESC''', ''') AS q WHERE program = "Harmattan" %s ''' % bugs_restrictions],
            },
            'bugs_activity': {
                'postfix': ['''
                LEFT JOIN bugs b ON a.bug_id = b.bug_id
                LEFT JOIN products AS pd ON b.product_id = pd.id
                LEFT JOIN classifications AS c ON (pd.classification_id = c.id)
                WHERE c.name = "Harmattan" AND a.attach_id IS NULL AND
                fieldid IN (53, 3, 14, 8, 10, 11, 12, 13, 42) %s
                ORDER BY bug_id, ts DESC''' % ba_restrictions],
            },
        }

        args = dict(config.parser.items('connection'))
        url = sa.engine.url.URL('mysql', **args)
        engine = sa.create_engine(url)
        engine.echo = config.run.debug_sql
        self._engine = engine
        
    def run(self):
        """
        
        """
        logging.info('Stage: dump')
        template = """%(prefix)s SELECT %(cols)s FROM %(table)s a %(postfix)s"""
        
        for name, cols in self._tables.iteritems():
            prefix = []
            postfix = []
                
            if name in self._tables_addons:
                addons = self._tables_addons
                if 'cols' in addons[name]:
                    cols.extend(addons[name]['cols'])
                if 'postfix' in addons[name]:
                    postfix.extend(addons[name]['postfix'])
                if 'prefix' in addons[name]:
                    prefix.extend(addons[name]['prefix'])
            sql = sa.text(template % dict(cols=', '.join(cols), table=name,
                postfix=' '.join(postfix), prefix=' '.join(prefix)))
            self._dump_query(self._engine.execute(sql), cols, name)
    
    def _dump_query(self, query, cols, table_name):
        """
        
        """
        cache = self.cache
        _cols = [x.split(' ')[-1] for x in cols]
        table = cache.make_table(_cols, table_name, cache.dumps)
        row = table.row
        for item in query:
            for i, value in enumerate(item):
                row[_cols[i]] = value
            row.append()
        table.flush()


bugzillametrics/cache.py
"""PyTables cache

"""

import logging

from tables import *

range = xrange


class Cache(object):
    """Responsible for managing dumps from Bugzilla and the rebuilt history.
    
    """
    def __init__(self, config):
        """
        
        """
        self.config = config
        self.mode = self.get_mode(config.run.start_at)
        file = openFile(config.workdir.cache_file, mode=self.mode,
            title='Cache')
        logging.info('Using cache file: %s', config.workdir.cache_file)
        self.file = file
        root = self.root = file.root
        # XXX: move to dump
        if not 'dumps' in root:
            dumps = file.createGroup('/', 'dumps', 'Database dumps')
        else:
            dumps = root.dumps
        self.dumps = dumps
        
    @staticmethod
    def get_mode(start_at):
        """Choose the most conservative file mode based on start stage.
        
        """
        if start_at == 'count':
            mode = 'r'
        elif start_at == 'rebuild':
            mode = 'a'
        elif start_at == 'dump':
            mode = 'w'
        return mode
    
    def make_table(self, cols, name, context):
        """Make a "dumb" table to save dumps in, i.e. a table with every
        column with the same varchar(255) (in SQL terms) type. Performance 
        could be improved by inspecting the SQL table definition and using 
        corresponding HDFS5 types, but it's fast now anyway.
        
        """
        result = {}
        for i in range(0, len(cols)):
            result[cols[i]] = StringCol(255, pos=i)
        table = type(name, (IsDescription,), result)
        return self.file.createTable(context, name, table)
        
        
class Coord(IsDescription):
    """PyTables model for rebuilt history coordinates. Comments show the 
    corresponding SQL types used by Bugzilla. Performance could be improved by 
    using EnumCol type for columns with fixed contents.
    
    A slight problem is that timestamps are saved as integers, so they need
    to be converted to datetime objects when in use, and this conversion
    should be automated somehow by using a custom type for this model.
    
    """
    # bugs.bug_id mediumint(9)
    id = IntCol(4, pos=0)
    # bugs.keywords mediumtext
    keywords = StringCol(128, pos=1)
    # bugs.bug_status varchar(64)
    status = StringCol(16, pos=2)
    #    status = EnumCol(Enum(Retriever.STATUS_TYPES), 'creation', base=Int32Atom())
    # bugs.resolution varchar(64)
    resolution = StringCol(64, pos=3)
    # bugs.bug_severity varchar(64)
    severity = StringCol(64, pos=4)
    # bugs.creation_ts datetime
    created = IntCol(8, pos=5)
    # bugs.priority varchar(64)
    priority = StringCol(64, pos=6)
    # classifications.name varchar(64)
    program = StringCol(64, pos=7)
    # products.name varchar(64)
    component = StringCol(64, pos=8)
    # components.name varchar(64)
    subcomponent = StringCol(64, pos=9)
    # bugs.delta_ts datetime
    ts = IntCol(8, pos=10)
    # bugs.creation_ts datetime
    creation_ts = IntCol(8, pos=11)
    # flags info varchar(64)
    flags = StringCol(128, pos=12)
    # changed_fields, stores the field that changed varchar(64)
    changed_fields = StringCol(128, pos=13)
    # removed stores what was removed varchar(64)
    removed = StringCol(256, pos=14)


bugzillametrics/config.py
"""Global config

"""

import io
import logging
from ConfigParser import ConfigParser, NoOptionError
from optparse import OptionParser



class Config(object):
    """Manage defaults and allow natural naming access to config, e.g.
    >>> value = config.group.key # natural naming
    instead of
    >>> value = config.get('group', 'key')
    
    """
    config_file = "/etc/bugzillametrics/bugzillametrics.cfg"
    def __init__(self, parser=ConfigParser()):
        """Set up defaults and user config.
        
        """
        # Read defaults
        defaults = open(self.config_file, 'r')
        parser.readfp(defaults)
        # Read user config if provided
        self.parser = parser

    def __getattr__(self, section_name):
        """Access options using a natural naming scheme.
        
        """
        parser = self.parser
        class Section(object):
            def __getattr__(self, option_name):
                return parser.get(section_name, option_name)
            def __setattr__(self, option_name):
                return parser.set(section_name, option_name)
        return Section()

    def options(self, section_name):
        return self.parser.options(section_name)

    def get_granularity(self):
        '''Returns granularity from config.
           Defaults to weeks if granularity is not present.'''
        days = {'days':1, 'weeks':7}
        try:
            granularity = self.parser.get('run','granularity')
        except NoOptionError:
            granularity = 'weeks'
        return (granularity, days[granularity])

    def update(self, args):
        "Handles command line options"
        optparser = OptionParser()

        optparser.add_option('-c', '--config', help='Pass an additional config file. Options specified in this file will override the defaults', metavar='file', type='string')
        optparser.add_option('-H', '--host', help='Database hostname', metavar='host', type='string')
        optparser.add_option('-u', '--user', help='Databse username', metavar='username', type='string')
        optparser.add_option('-p', '--passwd', help='Database password', metavar='passwd', type='string')
        optparser.add_option('-d', '--db', help='Database name', metavar='dbname', type='string')
        optparser.add_option('-o', '--outdir', help='Output dir (where the results will be stored)', metavar='outdir', type='string')
        optparser.add_option('-v', '--verbose', help='Increase log verbosity (debug mode)', action="store_true", default=False)
        optparser.add_option('--start', help='Phase to start from (dump/rebuild/count)', metavar='start_at', type='string')
        optparser.add_option('--stop', help='Phase to stop at (dump/rebuild/count)', metavar='stop_at', type='string')
        optparser.add_option('--today', help='Override current date', metavar='YYYY-MM-DD', type='string')
        optparser.add_option('--restrict', help='Restrict the dump to these bugs (comma separated list)', metavar='<bug1>,<bug2>,...', type='string')

        (options, _) = optparser.parse_args(args)
        # just a shorthand
        db = 'connection'
        phases = ['dump', 'rebuild', 'count']
        if options.config:
            # if additional config file is passed then use it
            self.parser.read(options.config)
        else:
            # Otherwise use the options passed if any
            if options.host:
                self.parser.set(db, 'host', options.host)
            if options.user:
                self.parser.set(db, 'user', options.user)
            if options.passwd:
                self.parser.set(db, 'password', options.passwd)
            if options.db:
                self.parser.set(db, 'database', options.db)
            if options.outdir:
                self.parser.set('workdir', 'path', options.outdir)
            if options.verbose:
                self.parser.set('workdir', 'log_level', 'debug')
            if options.start and options.start.lower() in phases:
                self.parser.set('run', 'start_at', options.start)
            if options.stop and options.stop.lower() in phases:
                self.parser.set('run', 'stop_at', options.stop)
            if options.today:
                self.parser.set('run', 'today', options.today)
            if options.restrict:
                self.parser.set('run', 'restrict', options.restrict)

            


bugzillametrics/rebuild.py
"""

"""
import logging
from collections import defaultdict
from pprint import pprint as pp
from datetime import date,datetime

from cache import Coord
from dump import Dump

class Rebuild(object):
    """
    
    """
    def __init__(self, config, cache):
        """
        
        """
        self.config = config
        self.cache = cache
        # TODO: don't access cache.file directly
        self.root = self.cache.file.root
        self.additive_fields = ['flags', 'keywords']
        self.closed_states = ['resolved', 'verified', 'closed', 'pre_verified', 'released']
        
    def _remove_history(self):
        """
        
        """
        root = self.root
        if 'history' in root:
            root.history.remove()
        
    def _reset_history(self):
        """
        
        """
        self._remove_history()
        self.root.history = self.cache.file.createTable('/', 'history', Coord, 'Bug coordinates')

    def set_next_events(self, ts, group, field, added, will_change):
        '''set next events in case they were not like they should have been'''
        gk = group.keys()
        gk.sort(cmp=lambda x,y:y-x)
        rem = 'removed'
        ch_flds = 'changed_fields'
        for next_ts in gk:
            if next_ts >= ts:
                event = group[next_ts]
                if field in self.additive_fields:
                    if added != '':
                        flags_to_add = added.split(',')
                        for fgta in flags_to_add:
                            if will_change[field].has_key(fgta):
                                #print "(found key %s for will_change): at %s, %s will change on %s" % (field, datetime.fromtimestamp(next_ts), field, datetime.fromtimestamp(will_change[field][fgta]))
                                if next_ts == will_change[field][fgta]:
                                    #print "next_ts: %s, will_change/%s/%s: %s" % (datetime.fromtimestamp(next_ts), field, fgta, datetime.fromtimestamp(will_change[field][fgta]))
                                    #print '===trying to set %s as %s for ts:%s' % (rem, fgta.lower(), datetime.fromtimestamp(next_ts))
                                    if event.has_key(rem):
                                        if event[rem] != '':
                                            removed = dict([x.split('>>') for x in event[rem].split('||')])
                                            self.update_this(field, ['', fgta.lower()], removed, will_change, next_ts)
                                            event[rem] = '||'.join(['%s>>%s' % (x, removed[x]) for x in removed])
                                    else:
                                        event[rem] = '%s>>%s' % (field,added.lower())

                                    if event.has_key(ch_flds):
                                        changed_fields = event[ch_flds].split('||')
                                        changed_fields.append(field)
                                        event[ch_flds] = '||'.join(changed_fields)
                                    else:
                                        event[ch_flds] = field
                                        
                                if next_ts < will_change[field][fgta]:
                                    self.update_this(field, ['', fgta.lower()], event, will_change, next_ts)
                            else:
                                #print '(no key %s for will_change): trying to set %s as %s for ts:%s' % (field, field, fgta.lower(), datetime.fromtimestamp(next_ts))
                                self.update_this(field, ['', fgta.lower()], event, will_change, next_ts)
                                
                else:
                    if will_change.has_key(field):
                        #print "(found key %s for will_change): at %s, %s will change on %s" % (field, datetime.fromtimestamp(next_ts), field, datetime.fromtimestamp(will_change[field]))
                        if next_ts < will_change[field]:
                            self.update_this(field, ['', added.lower()], event, will_change, next_ts)
                        if next_ts == will_change[field]:
                            #print '===trying to set %s as %s for ts:%s' % (rem, added.lower(), datetime.fromtimestamp(next_ts))
                            if event.has_key(rem):
                                if event[rem] != '':
                                    removed = dict([x.split('>>') for x in event[rem].split('||')])
                                    removed[field] = added.lower()
                                    event[rem] = '||'.join(['%s>>%s' % (x, removed[x]) for x in removed])
                            else:
                                event[rem] = '%s>>%s' % (field,added.lower())
                            if event.has_key(ch_flds):
                                changed_fields = event[ch_flds].split('||')
                                changed_fields.append(field)
                                event[ch_flds] = '||'.join(changed_fields)
                            else:
                                event[ch_flds] = field
                    else:
                        #print '(no key %s for will_change): trying to set %s as %s for ts:%s' % (field, field, added.lower(), datetime.fromtimestamp(next_ts))
                        event[field] = added.lower()
                        will_change[field] = ts

    def update_this(self, field, changes, event, will_change, ts):
        """updates this to event[field]
           Takes care of additive fields, like keywords and flags, properly
           
           changes is a list of [to_be_removed, to_be_added] fields"""
        if field in self.additive_fields:
            if changes[1] != '':
                add_fields = changes[1].split(',')
                #print "adding %s to %s" % (','.join(add_fields), field)
                if event.has_key(field):
                    if event[field] != '':
                        cur_flags = event[field].split(',')
                        event[field] = ','.join(list(set.union(set(add_fields),set(cur_flags))))
                    else:
                        event[field] = ','.join(add_fields)
                else:
                    event[field] = ','.join(add_fields)
                #print field,':',event[field]
                for wc in add_fields:
                    #if will_change[field].has_key(wc):
                        #print "[update_add] updating will_change for %s/%s to %s (was %s)" % (field,wc,datetime.fromtimestamp(ts), datetime.fromtimestamp(will_change[field][wc]))
                    #else:   
                        #print "[update_add] updating will_change for %s/%s to %s (wasn't there) " % (field,wc,datetime.fromtimestamp(ts))
                    will_change[field][wc] = ts
            if changes[0] != '':
                rem_fields = changes[0].split(',')
                #print "removing %s from %s" % (','.join(rem_fields), field)
                if event.has_key(field):
                    if event[field] != '':
                        cur_flags = event[field].split(',')
                        #print 'current flags: ', cur_flags,' - rem_fields: ', rem_fields
                        event[field] = ','.join(list(set.difference(set(cur_flags),set(rem_fields))))
                        #print ' is ', event[field]

                    #else:
                        #print "event empty, can't remove anything from there"
                        #print "Need to set it also for all the other next events!!"
                #else:
                    ##print "event empty, can't remove anything from there"
                    #print "Need to set it also for all the other next events!!"
        elif not field in self.additive_fields:
            #print "Setting %s for %s" % (changes[1], field)
            event[field] = changes[1]
            will_change[field] = ts

    def run(self):
        """
        
        """
        logging.info('Stage: rebuild history')
        self._reset_history()
        root = self.root
        activities = root.dumps.bugs_activity
        activities_fields = activities.colnames
        creations = root.dumps.bugs
        creations_fields = creations.colnames
        field_map = {
            3: 'component',
            8: 'status',
            10: 'keywords',
            11: 'resolution',
            12: 'severity',
            13: 'priority',
            14: 'subcomponent',
            53: 'program',
            42: 'flags',
        }
        # if True changing component resets priority
        self.component_reset_status = True

        n = len(activities)
        i = 0
        creation_ids = list(frozenset([bug['id'] for bug in creations]))
        n_bugs = len(creation_ids)
        print "Found %s bugs" % n_bugs
        creation_ids.sort()
        lowercase_fields = frozenset(field_map.values())
        rem = 'removed'
        ch_flds = 'changed_fields'
        prio = 'priority'
        stat = 'status'
        processed_bugs = 0
        for bug in creations:
            # The theory is that each given field has its status until it's changed
            group = {}
            bug_id = bug['id']
            prc = float(processed_bugs)/float(n_bugs) * 100
            print 'Processed %.2f%% bugs' % prc,
            print '\r',
            # Map bug fields to field names
            bug = dict(zip(creations_fields, bug[:]))
            # TODO: use datetime objects
            ts = int(bug['ts'])
            bug['ts'] = ts
            # TODO: do this at dump stage
            will_change = {}
            for field in lowercase_fields:
                if bug.has_key(field):
                    bug[field] = bug[field].lower()
                else:
                    bug[field] = ''
                if field in self.additive_fields:
                    will_change[field] = {}
                else:
                    will_change.setdefault(field, int(ts)+1)
            # Buffer creation coordinate
            group[ts] = bug.copy()
            next_ts = ts
            prev_event = {}
                
            # Look for matching activity coordinates
            while i < n:
                # Map coordinate fields to field names
                coord_id = activities[i]['bug_id']
                if not coord_id == bug_id:
                    # Reached end of coordinates for current bug
                    # Go to next bug with the same coordinate until ids match
                    break
                if not coord_id in creation_ids:
                    # Skip orphan coordinates (should never happen)
                    logging.warning('Orphan coordinate: %s' % (coord_id,))
                    i += 1
                    continue
                coord = dict(zip(activities_fields, activities[i]))
                ts = int(coord['ts'])
                # Merge coordinate with previous state
                try:
                    # field is the field that changed
                    # all the rest will stay the same 
                    field = field_map[int(coord['fieldid'])]
                except ValueError, e:
                    pp(coord['fieldid'])
                added = coord['added'].lower()
                removed = coord[rem].lower()
                if not added and not removed:
                    # Skipping stuff that didn't have activity (flags handling)
                    i += 1
                    continue
                # this keeps a record of last ts when some field got changed
                # it is there to cope with changes that don't get written to the activities
                dts= datetime.fromtimestamp(ts)
                if field in lowercase_fields:
                    if field in self.additive_fields:
                        removed = ','.join([x.strip() for x in removed.split(',')])
                        added = ','.join([x.strip() for x in added.split(',')])
                    #print '[%s] field: %s, added: %s, removed: %s' % (datetime.fromtimestamp(ts), field, added, removed)
                    if ts != next_ts:
                        next_event = group[next_ts]
                        # this is a new coordinate in time
                        # copy the stuff from before (since we go backward, before is next)
                        cur_event = group[ts] = next_event.copy()
                        for ks in prev_event.keys():
                            if cur_event[ks] != prev_event[ks]:
                                # set what we know are the next keys
                                self.update_this(ks, prev_event[ks], cur_event, will_change, ts)
                            del prev_event[ks]
                        # set new datetime
                        cur_event['ts'] = ts
                        # set field that changed/was added
                        if cur_event[field] != added:
                            # we need to set it also for later as it is not like this
                            self.set_next_events(ts, group, field, added, will_change)
                        # initialize (new event) and record what changed
                        cur_event[ch_flds] = field
                        cur_event[rem] = '%s>>%s' % (field, removed)
                        # initialize next event dict
                        # remember what was the next_ts
                        next_ts = ts
                    else:
                        # we have already this time coordinate recorded
                        # added should be already there since we are going
                        # in reverse order
                        cur_event = group[ts]
                        if not added.lower() == cur_event[field].lower():
                            # WTF we didn't know this status already
                            # Let's go through all the existing bugs 
                            # that have ts higher than this and set it
                            self.set_next_events(ts, group, field, added, will_change)
                        changed_fields = []
                        if cur_event.has_key(ch_flds):
                            changed_fields = cur_event[ch_flds].split('||')
                        changed_fields.append(field)
                        cur_event[ch_flds] = '||'.join(changed_fields)
                        rem_fields = []
                        if cur_event.has_key(rem):
                            rem_fields = cur_event[rem].split('||')
                        rem_fields.append('%s>>%s' % (field, removed))
                        cur_event[rem] = '||'.join(rem_fields)
                        if field == 'component' and self.component_reset_status \
                        and prio not in changed_fields \
                        and cur_event['status'] not in self.closed_states:
                            changed_fields.append('priority')
                            cur_event[ch_flds] = '||'.join(changed_fields)
                            rem_fields.append('priority>>%s' % (cur_event[prio]))
                            cur_event[rem] = '||'.join(rem_fields)
                            prev_event[prio] = ['', cur_event[prio]]
                            will_change[prio] = ts
                            cur_event[prio] = 'unspecified'
                    # and remember that prev_event has field as removed 
                    prev_event[field] = [added,removed]
                i += 1
            # Ok now we are out of bugs_activity it means we are at the bug creation
            ts = int(bug['creation_ts'])
            first_event = group[ts] = group[next_ts].copy()
            for ks in prev_event.keys():
                if ks in self.additive_fields and ts == next_ts:
                    # This happened at the same time of the creation
                    # So we need to actually add this to the first_event, not remove it
                    #print "creation: %s , next_ts: %s" % (datetime.fromtimestamp(ts), datetime.fromtimestamp(next_ts))
                    #print "%s: %s" % (ks,prev_event[ks])
                    prev_event[ks].reverse()
                # set what we know are the next keys
                self.update_this(ks, prev_event[ks], first_event, will_change, ts)
            first_event[ch_flds] = ''
            first_event[rem] = ''
            first_event['ts'] = ts
            bug_history = []
            gk = group.keys()
            gk.sort(cmp=lambda x,y:int(y)-int(x))
            for timestamp in gk:
                bug_history.append(group[timestamp])
            # Push the complete buffered group of bug coordinates
            self.commit(bug_history)
        # Clean up
        # TODO: don't use cache.file?
            processed_bugs += 1
        self.cache.file.flush()
    def commit(self, group):
        """
        
        """
        if self.cache.mode == 'r':
            return
        row = self.root.history.row
        for bug in group:
            for k, v in bug.iteritems():
                row[k] = v
            row.append()
    
#    def reset(self):
#        if 'history' in root:
#            root.history.remove()
#        if not 'history' in root:
#            file.createTable('/', 'history', Coord, 'Bug coordinates')
